fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/aggregate/AggregateCursor.java

```diff
@@ -57,77 +58,104 @@ public class AggregateCursor<M extends Message> implements RecordCursor<QueryRes
+            // either innerResult.hasNext() = false; or groupBreak = true
+            if (Verify.verifyNotNull(previousResult).hasNext()) {
+                // in this case groupBreak = true, return aggregated result and continuation
+                RecordCursorContinuation continuation = Verify.verifyNotNull(previousValidResult).getContinuation();
+                /*
+                * Update the previousValidResult to the next continuation even though it hasn't been returned. This is to return the correct continuation when there are single-element groups.
+                * Below is an example that shows how continuation(previousValidResult) moves:
+                * Initial: previousResult = null, previousValidResult = null
+                row0      groupKey0      groupBreak = False         previousValidResult = row0  previousResult = row0
+                row1      groupKey0      groupBreak = False         previousValidResult = row1  previousResult = row1
+                row2      groupKey1      groupBreak = True          previousValidResult = row1  previousResult = row2
+                * returns result (groupKey0, continuation = row1), and set previousValidResult = row2
+                *
+                * Now there are 2 scenarios, 1) the current iteration continues; 2) the current iteration stops
+                * In scenario 1, the iteration continues, it gets to row3:
+                row3      groupKey2      groupBreak = True          previousValidResult = row2  previousResult = row3
+                * returns result (groupKey1, continuation = row2), and set previousValidResult = row3
+                *
+                * In scenario 2, a new iteration starts from row2 (because the last returned continuation = row1), and set initial previousResult = null, previousValidResult = null:
+                row2      groupKey1      groupBreak = False         previousValidResult = row2  previousResult = row2
+                * (Note that because a new iteration starts, groupBreak = False for row2.)
+                row3      groupKey2      groupBreak = True          previousValidResult = row2  previousResult = row3
+                * returns result (groupKey1, continuation = row2), and set previousValidResult = row3
+                *
+                * Both scenarios returns the correct result, and continuation are both set to row3 in the end, row2 is scanned twice if a new iteration starts.
+                */
+                previousValidResult = previousResult;
+                return RecordCursorResult.withNextValue(QueryResult.ofComputed(streamGrouping.getCompletedGroupResult()), continuation);
+            } else {
+                if (Verify.verifyNotNull(previousResult).getNoNextReason() == NoNextReason.SOURCE_EXHAUSTED) {
+                    if (previousValidResult == null) {
+                        return RecordCursorResult.exhausted();
```

Comment:
Does this need to check to insert the empty result if `isCreateDefaultOnEmpty`? It feels like that information is being lost here.

Perhaps it would be simpler if there were fewer moving parts here. If we wait until 4.2, we could drop support entirely for `isCreateDefaultOnEmpty` (see #3107) which may simplify this somewhat.